openshift/WFLY-15421_OIDC_for_WildFly_s2i_v2.adoc

```diff
@@ -0,0 +1,298 @@
+For example, `provider-url: ${env.OIDC_URL}` would imply that the env variable `OIDC_URL=http://keycloak/auth/realms/WildFly` is set.
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== User CLI scripts called at packaging time
+
+When using the https://issues.redhat.com/browse/WFLY-14934[WildFly Maven plugin] to package an application, a script can be called to configure the OIDC subsystem for the deployed application.
+
+This support works out of the box and doesn’t require any addition to the WildFly s2i image. It applies to any kind of OIDC provider.
+
+This implies a set of WildFly CLI management operations to be called at packaging time.
+
+For example:
+`/subsystem=elytron-oidc-client/secure-deployment=simple-webapp.war:add(client-id=simple-webapp, public-client=true, provider-url=${OIDC_URL}, ssl-required=EXTERNAL)`
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== Environment variable based configuration
+
+This is the main topic of this proposal.
+
+A CLI script is generated at boot time to handle the OIDC subsystem configuration based on discovered 
+deployments (actually deployed and found in content repository or located in `standalone/deployments` directory).
+In addition there is an automatic creation and registration of the clients in the authorization server.
+
+In order for war deployments to be discovered they must have the `OIDC` auth-method in web.xml. 
+Deployments that contain a `WEB-INF/oidc.json` descriptor file are ignored.
+
+NB: A single provider can be configured for a running server. The discovered deployments are all registered inside the same OIDC provider.
+
+We are offering a new set of env variables `OIDC_*` to be used to configure the subsystem, enable the discovery of deployments and registration of 
+clients into the authorization server.
+
+The existing `SSO_*` env variables used for OIDC are usable for Keycloak OIDC integration. 
+It helps transition smoothly to use the native OIDC support. 
+This proposal defines a mapping between the `SSO_*` and `OIDC_*` env variables.
+The `SSO_*` env variables are deprecated.
+
+The env variable `OIDC_PROVIDER_NAME=<name>` MUST be set when using `OIDC_*` env variables. When using `SSO_*` env variables 
+the provider name `keycloak` is internally used.
+
+A mix of `SSO_*` and `OIDC_*` env variables, although not recommended, would work, 
+new env variables will only be added in the `OIDC_*` namespace. We have a first example `OIDC_PROVIDER_SSL_REQUIRED` 
+that would be new without a mapping in the `SSO_*` namespace.
+
+Example of configuration using new OIDC env variables for Keycloak:
+
+```
+OIDC_PROVIDER_NAME=keycloak
+OIDC_PROVIDER_URL=http://keycloak/auth/realms/Wildfly
+OIDC_USER_NAME=foo
+OIDC_USER_PASSWORD=foo
```

Comment:
Are `OIDC_USER_NAME` and `OIDC_USER_PASSWORD` the client id and client secret?

`OIDC_CLIENT_ID` and `OIDC_CLIENT_SECRET` might be better variable names.

Reply:
Actually, I also see `OIDC_SECURE_DEPLOYMENT_SECRET` below.

How are the `OIDC_USER_NAME`, `OIDC_USER_PASSWORD`, and `OIDC_SECURE_DEPLOYMENT_SECRET` variables used.

Reply:
The NAME and PASSWORD are credetials used to retrieve the token from the keycloak server (token required to then create a new remote client). These pieces of information are not in the subsystem configuration.
The SECRET is the actual secret set in the secure deployment configuration. This piece of information is also sent to the server when creating/registering the client in the remote keycloak server.